File: solutions/python/meetup/meetup.py
import datetime
import enum


class Weekday(enum.IntEnum):
  Monday = 0
  Tuesday = 1
  Wednesday = 2
  Thursday = 3
  Friday = 4
  Saturday = 5
  Sunday = 6


class MeetupDayException(Exception):
  pass


OFFSET = {'1st': 1, '2nd': 8, '3rd': 15, '4th': 22, '5th': 29, 'teenth': 13}


def meetup(year, month, week, day_of_week):
  
  if week in OFFSET:
    try:
      d = datetime.date(year, month, OFFSET[week])
    except ValueError:
      raise MeetupDayException('Invalid')
  elif week == 'last':
    month += 1
    if month > 12:
      year += month // 12
      month %= 12
    d = datetime.date(year, month, 1) - datetime.timedelta(7)
  else:
    raise MeetupDayException('Invalid')

  day_of_week = Weekday[day_of_week].value
  dow_shift = datetime.timedelta((day_of_week - d.weekday()) % 7)
  return d + dow_shift


# vim:ts=2:sw=2:expandtab

File: solutions/ruby/meetup/meetup.rb
require 'date'

class Meetup
  WDAYS = Date::DAYNAMES.map { |name| name.downcase.to_sym }

  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(dow, week)
    if week == :last
      d = Date.new(@year, @month, 1).next_month.prev_day
      shift = -1
    else
      offset = { first: 1, second: 8, third: 15, fourth: 22, fifth: 29, teenth: 13 }[week]
      d = Date.new(@year, @month, offset)
      shift = +1
    end

    d += shift while WDAYS[d.wday] != dow
    d
  end
end
